### ISSUE TITLE
Remove no_cuda_on_cpu_tap tag and fix tests

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#46425]
This is using TF 2.4.0

**Describe the problem**

When building and running tests I stumbled over the seemingly required test tag `no_cuda_on_cpu_tap` which is only documented in one place:
```
    # Running cuda on cpu will trigger tests guarded by GOOGLE_CUDA but NCHW
    # won't be available, which result in test failures. So disable that.
```

This sounds like a workaround for a problem when writing the tests. The most obvious incarnation for this is e.g. `//tensorflow/core/common_runtime:ring_reducer_test`. That is added with `tf_cuda_cc_test` which adds 2 tests: One with `_gpu` suffix gpu tags and one without. However as the test code uses `GOOGLE_CUDA` to decide whether to test on CPU or GPU the CPU test version cannot be run when TF is built with CUDA, which is (AFAIK) the regular case.

Hence my suggestion would be to "fix" the tests to run only on GPU if it is tagged with `gpu` and pass that to the test code as a define (e.g. `TF_TEST_USE_GPU`)

**Comment by SuryanarayanaY** [comment]
Hi @Flamefire,

Apologies for the long delay.Assignee seems not working with TF team now and hence not sure of the status yet.

Could you please let us know is this still an issue in latest TF versions ?

**Comment by Flamefire** [comment]
Yes the same issue as described is still present in latest master.

See https://github.com/tensorflow/tensorflow/blob/a2ee21765553ae782439d833b7130dcba5897b6e/tensorflow/core/grappler/optimizers/BUILD#L137-L139 for the comment on the tag, https://github.com/tensorflow/tensorflow/blob/a2ee21765553ae782439d833b7130dcba5897b6e/tensorflow/core/common_runtime/BUILD#L2512-L2513 for the definition of the ring_reducer_test and the [erronous check for `GOOGLE_CUDA`](https://github.com/tensorflow/tensorflow/blob/a2ee21765553ae782439d833b7130dcba5897b6e/tensorflow/core/common_runtime/ring_reducer_test.cc#L403) in that test and https://github.com/tensorflow/tensorflow/blob/a2ee21765553ae782439d833b7130dcba5897b6e/tensorflow/tensorflow.bzl#L453 for the definition of `GOOGLE_CUDA` depending on if TF is built with CUDA support

**Comment by Flamefire** [comment]
@Venkat6871 I double checked everything and the code is unchaged in current master, so the issue is still present.

**Comment by Kayyuri** [comment]
Hi @Flamefire,
Thank you so much for filing this issue! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!
